README.md
# Search Management Map Asset API

A C library for accessing the asset side of the API for [Search Management Map](https://github.com/canterburyairpatrol/search-management-map)

## Getting started
### Prerequisites

 * [libcurl](https://curl.haxx.se)
 * [libtidy](http://html-tidy.org)
 * [Jansson](http://www.digip.org/jansson/)

### Fetching and building

```
git clone https://github.com/canterbury-air-patrol/smm-asset-api.git
cd smm-asset-api
./autogen.sh
./configure -prefix=/usr
make
make install
```

### Using

Link your program against -lsmmasset or use pkg-config with the installed smm-asset.pc.

Basic API usage example
```
#include <smm-asset.h>

int main(int argc, char *argv[])
{
	smm_asset *assets;
	size_t assets_count;
	smm_asset asset = NULL;

	smm_connection conn = smm_asset_connect ("http://localhost/", "asset", "assetpassword");
	if (smm_asset_get_assets (conn, &assets, assets_count))
	{
		for (size_t i = 0; i < assets_count; i++)
		{
			if (strcmp (smm_asset_get_name (assets[i]), "my-asset") == 0)
			{
				asset = assets[i];
			}
		}
	}
	if (asset != NULL)
	{
		/* Report this assets position as 43 deg South, 172 deg East, 35 meters high, heading west, 3d fix */ 
		smm_asset_report_position (asset, -43, 172, 35, 270, 3);

		/* You can also get a search to conduct by */
		smm_search search;
		do {
			search = smm_asset_get_search (asset, -43, 172);
		} while (!smm_search_accept (search));
		/* All the waypoints for the search are accessible with */
		smm_waypoints waypoints;
		size_t waypoints_count;
		smm_search_waypoints_get (search, &waypoints, &waypoints_count);
		/* Iterate them the same as assets, then free with */
		smm_waypoints_free (waypoints, waypoints_count);
		/* Once the search is completed then */
		smm_search_completed (search);
		smm_search_destroy (search);
	}

	smm_asset_free_assets (assets, assets_count);
	smm_connection_close (conn);
}
```

## Authors
See the list of [contributors](https://github.com/canterbury-air-patrol/smm-asset-api/contributors).

## License
This project is licensed under GNU LGPLv2.1 see the [LICENSE.md](LICENSE.md) file for details.



debian/changelog
libsmm-asset (0.2.3) unstable; urgency=medium

  * Updated

 -- Scott Parlane <sjp@canterburyairpatrol.org>  Sun, 19 Jul 2020 00:00:00 +0000



src/smm-asset.c
/**
 * smm-asset.c, API functions for communicating with Search Management Map
 * to act as an Asset.
 *
 * Copyright 2019 Canterbury Air Patrol Incorporated
 *
 * This file is part of libsmm-asset
 *
 * libsmm-asset is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "smm-asset.h"
#include "smm-asset-internal.h"

#include <pthread.h>
#include <stdlib.h>
#include <string.h>

#include <jansson.h>

bool smm_debug = false;

void
smm_asset_debugging_set (bool debug)
{
	smm_debug = debug;
}

smm_connection
smm_asset_connect (const char *host, const char *user, const char *pass)
{
	smm_connection conn = calloc (1, sizeof (struct smm_connection_s));
	if (conn == NULL)
	{
		return NULL;
	}

	conn->host = strdup (host);
	conn->user = strdup (user);
	conn->pass = strdup (pass);
	pthread_mutex_init (&conn->lock, NULL);

	smm_connection_login (conn);

	return conn;
}

smm_connection_status
smm_asset_connection_get_state (smm_connection connection)
{
	if (connection == NULL)
	{
		return SMM_CONNECTION_UNKNOWN;
	}
	return connection->state;
}

void
smm_connection_close (smm_connection connection)
{
	if (connection != NULL)
	{
		free (connection->host);
		free (connection->user);
		free (connection->pass);
		free (connection->csrfmiddlewaretoken);
		curl_easy_cleanup (connection->curl);
		pthread_mutex_destroy(&connection->lock);
	}
	free (connection);
}

smm_asset
smm_asset_create (smm_connection conn, const char *name, const char *type, int asset_id, int asset_type_id)
{
	smm_asset asset = calloc (1, sizeof (struct smm_asset_s));
	asset->conn = conn;
	asset->name = strdup (name);
	asset->type = strdup (type);
	asset->asset_id = asset_id;
	asset->asset_type_id = asset_type_id;

	return asset;
}

bool
smm_asset_get_assets (smm_connection connection, smm_assets * assets, size_t * assets_count)
{
	struct buffer_s buf = { NULL, 0 };
	json_t *json_root;
	json_error_t json_error;

	struct smm_curl_res_s *res = smm_connection_curl_retrieve_url (connection, "/assets/mine/json/", NULL, to_buffer, &buf);
	if (res == NULL)
	{
		return false;
	}
	if (!(res->success && res->httpcode == 200))
	{
		smm_curl_res_free (res);
		return false;
	}
	smm_curl_res_free (res);

	/* Parse the assets */
	*assets_count = 0;
	*assets = NULL;

	json_root = json_loadb (buf.data, buf.bytes, 0, &json_error);
	if (json_root)
	{
		json_t *json_assets = json_object_get (json_root, "assets");
		if (json_assets)
		{
			size_t index;
			json_t *value;

			json_array_foreach (json_assets, index, value)
			{
				const char *key;
				json_t *val;
				int asset_id = -1;
				int asset_type_id = -1;
				const char *name = NULL;
				const char *type = NULL;
				json_object_foreach (value, key, val)
				{
					if (strcmp (key, "id") == 0)
					{
						asset_id = json_integer_value (val);
					}
					else if (strcmp (key, "type_id") == 0)
					{
						asset_type_id = json_integer_value (val);
					}
					else if (strcmp (key, "name") == 0)
					{
						name = json_string_value (val);
					}
					else if (strcmp (key, "type_name") == 0)
					{
						type = json_string_value (val);
					}
					else
					{
						if (json_is_integer (val))
						{
							printf ("%s = %lli\n", key, json_integer_value (val));
						}
						else if (json_is_string (val))
						{
							printf ("%s = %s\n", key, json_string_value (val));
						}
					}
				}
				*(assets_count) += 1;
				*assets = realloc (*assets, *assets_count * sizeof (smm_asset));
				(*assets)[(*assets_count) - 1] = smm_asset_create (connection, name, type, asset_id, asset_type_id);
			}
		}
		else
		{
			printf ("Didn't find assets\n");
		}
	}
	else
	{
		printf ("Error on line %i: %s\n", json_error.line, json_error.text);
	}

	json_decref (json_root);


	free (buf.data);
	return true;
}


void
smm_asset_free_asset (smm_asset asset)
{
	free (asset->name);
	free (asset->type);
	free (asset);
}

void
smm_asset_free_assets (smm_assets assets, size_t assets_count)
{
	for (size_t i = 0; i < assets_count; i++)
	{
		smm_asset_free_asset (assets[i]);
	}
	free (assets);
}

static int
smm_asset_get_asset_id (smm_asset asset)
{
	return asset->asset_id;
}

const char *
smm_asset_name (smm_asset asset)
{
	if (asset)
	{
		return asset->name;
	}
	return NULL;
}

const char *
smm_asset_type (smm_asset asset)
{
	if (asset)
	{
		return asset->type;
	}
	return NULL;
}

static bool
smm_asset_update_command (smm_asset asset, struct buffer_s *buf)
{
	json_t *json_root;
	json_error_t json_error;

	json_root = json_loadb (buf->data, buf->bytes, 0, &json_error);
	if (json_root)
	{
		const char *command = NULL;
		json_t *tmp;

		tmp = json_object_get (json_root, "action");
		if (tmp)
		{
			command = json_string_value (tmp);
		}

		if (strcmp (command, "GOTO") == 0)
		{
			/* Get lat and long as well */
			tmp = json_object_get (json_root, "latitude");
			if (tmp)
			{
				asset->last_command_lat = json_real_value (tmp);
			}
			tmp = json_object_get (json_root, "longitude");
			if (tmp)
			{
				asset->last_command_lon = json_real_value (tmp);
			}
			asset->last_command = SMM_COMMAND_GOTO;
		}
		else if (strcmp (command, "RON") == 0)
		{
			asset->last_command = SMM_COMMAND_CONTINUE;
		}
		else if (strcmp (command, "RTL") == 0)
		{
			asset->last_command = SMM_COMMAND_RTL;
		}
		else if (strcmp (command, "CIR") == 0)
		{
			asset->last_command = SMM_COMMAND_CIRCLE;
		}
		else if (strcmp (command, "AS") == 0)
		{
			asset->last_command = SMM_COMMAND_ABANDON_SEARCH;
		}
		else if (strcmp (command, "MC") == 0)
		{
			asset->last_command = SMM_COMMAND_MISSION_COMPLETE;
		}
		else
		{
			asset->last_command = SMM_COMMAND_UNKNOWN;
		}


		json_decref (json_root);
	}
	else
	{
		printf ("Error on line %i: %s\n", json_error.line, json_error.text);
		asset->last_command = SMM_COMMAND_UNKNOWN;
	}

	return true;
}


smm_asset_command
smm_asset_last_command (smm_asset asset)
{
	return asset->last_command;
}

bool
smm_asset_last_goto_pos (smm_asset asset, double *lat, double *lon)
{
	if (asset->last_command != SMM_COMMAND_GOTO)
	{
		return false;
	}
	if (lat == NULL || lon == NULL)
	{
		return false;
	}
	*lat = asset->last_command_lat;
	*lon = asset->last_command_lon;
	return true;
}


bool
smm_asset_report_position (smm_asset asset, double latitude, double longitude, unsigned int altitude, uint16_t bearing, uint8_t fix)
{
	struct buffer_s buf = { NULL, 0 };

	char *page = NULL;
	if (asprintf (&page, "/data/assets/%s/position/add/?lat=%lf&lon=%lf&alt=%u&bearing=%u&fix=%u", asset->name, latitude, longitude, altitude, bearing, fix)
	    < 0)
	{
		return false;
	}

	struct smm_curl_res_s *res = smm_connection_curl_retrieve_url (asset->conn, page, NULL, to_buffer, &buf);
	if (res == NULL)
	{
		free (page);
		return false;
	}
	if (!(res->success && res->httpcode == 200))
	{
		smm_curl_res_free (res);
		free (page);
		return false;
	}

	free (page);

	/* if json data was returned, update the current action */
	if (res->content_type != NULL && strcmp (res->content_type, "application/json") == 0)
	{
		smm_asset_update_command (asset, &buf);
	}
	else
	{
		if (buf.data && strncmp (buf.data, "Continue", buf.bytes) == 0)
		{
			asset->last_command = SMM_COMMAND_CONTINUE;
		}
		else
		{
			asset->last_command = SMM_COMMAND_NONE;
		}
	}

	free (buf.data);

	smm_curl_res_free (res);

	return true;
}

static smm_search
smm_search_create (smm_asset asset, const char *url, uint64_t length, uint64_t distance, uint64_t sweep_width)
{
	smm_search search = calloc (1, sizeof (struct smm_search_s));
	search->asset = asset;
	search->url = strdup (url);
	search->length = length;
	search->distance = distance;
	search->sweep_width = sweep_width;

	return search;
}

uint64_t
smm_search_distance (smm_search search)
{
	if (search)
	{
		return search->distance;
	}
	return 0;
}

uint64_t
smm_search_length (smm_search search)
{
	if (search)
	{
		return search->length;
	}
	return 0;
}

uint64_t
smm_search_sweep_width(smm_search search)
{
	if (search)
	{
		return search->sweep_width;
	}
	return 0;
}

void
smm_search_destroy (smm_search search)
{
	if (search)
	{
		free (search->url);
		free (search);
	}
}

static smm_waypoint
smm_waypoint_create (float lat, float lon)
{
	smm_waypoint wp = calloc (1, sizeof (struct smm_waypoint_s));
	wp->lat = lat;
	wp->lon = lon;
	return wp;
}

static void
smm_waypoint_free (smm_waypoint waypoint)
{
	free (waypoint);
}


bool
smm_search_get_waypoints (smm_search search, smm_waypoints * waypoints, size_t * waypoints_count)
{
	struct buffer_s buf = { NULL, 0 };
	json_t *json_root;
	json_error_t json_error;

	struct smm_curl_res_s *res = smm_connection_curl_retrieve_url (search->asset->conn, search->url, NULL, to_buffer, &buf);

	if (res == NULL)
	{
		return false;
	}
	else if (!(res->success && res->httpcode == 200))
	{
		/* Login, try again */
		smm_curl_res_free (res);
		return false;
	}

	smm_curl_res_free (res);


	/* Parse the assets */
	*waypoints_count = 0;
	*waypoints = NULL;

	json_root = json_loadb (buf.data, buf.bytes, 0, &json_error);
	if (json_root)
	{
		json_t *json_features = json_object_get (json_root, "features");
		if (json_features)
		{
			if (json_array_size (json_features) == 1)
			{

				printf ("Parsing waypoints\n");
				size_t index;
				json_t *value;

				json_t *json_search = json_array_get (json_features, 0);
				if (json_search == NULL)
				{
					printf ("No json_search :(\n");
				}
				json_t *json_geometry = json_object_get (json_search, "geometry");
				if (json_geometry == NULL)
				{
					printf ("No json_geometry \n");
				}
				json_t *json_coords = json_object_get (json_geometry, "coordinates");
				if (json_coords == NULL)
				{
					printf ("No json_coords\n");
				}
				json_array_foreach (json_coords, index, value)
				{
					float lat = 0.0;
					float lon = 0.0;
					json_t *json_lat = json_array_get (value, 1);
					json_t *json_lon = json_array_get (value, 0);
					lat = json_real_value (json_lat);
					lon = json_real_value (json_lon);
					*(waypoints_count) += 1;
					*waypoints = realloc (*waypoints, *waypoints_count * sizeof (smm_waypoint));
					(*waypoints)[(*waypoints_count) - 1] = smm_waypoint_create (lat, lon);
				}
			}
			else
			{
				printf ("array size != 1 (%zi)", json_array_size (json_features));
			}
		}
		else
		{
			printf ("Didn't find waypoints\n");
		}
	}
	else
	{
		printf ("Error on line %i: %s\n", json_error.line, json_error.text);
	}

	json_decref (json_root);

	free (buf.data);

	return true;
}

static bool
smm_search_action (smm_search search, const char *action)
{
	char *action_page = NULL;
	struct buffer_s buf = { NULL, 0 };

	{
		char *tmp = strdup (search->url);
		char *json_str = strstr (tmp, "/json/");
		if (json_str)
		{
			*json_str = '\0';
			if (asprintf (&action_page, "%s/%s/?asset_id=%i", tmp, action, smm_asset_get_asset_id (search->asset)) < 0)
			{
				free (tmp);
				return false;
			}
		}
		free (tmp);
	}
	if (action_page == NULL)
	{
		return false;
	}

	struct smm_curl_res_s *res = smm_connection_curl_retrieve_url (search->asset->conn, action_page, NULL, to_buffer, &buf);
	if (res == NULL)
	{
		return false;
	}
	else if (!(res->success && res->httpcode == 200))
	{
		smm_curl_res_free (res);
		return false;
	}

	smm_curl_res_free (res);
	free (action_page);
	action_page = NULL;

	free (buf.data);

	return true;
}

bool
smm_search_accept (smm_search search)
{
	return smm_search_action (search, "begin");
}

bool
smm_search_complete (smm_search search)
{
	return smm_search_action (search, "finished");
}

void
smm_waypoints_free (smm_waypoints waypoints, size_t waypoints_count)
{
	for (size_t i = 0; i < waypoints_count; i++)
	{
		smm_waypoint_free (waypoints[i]);
	}
	free (waypoints);
}

smm_search
smm_asset_get_search (smm_asset asset, double latitude, double longitude)
{
	smm_search search = NULL;
	struct buffer_s buf = { NULL, 0 };

	char *page = NULL;
	if (asprintf (&page, "/search/find/closest/?asset_id=%i&latitude=%lf&longitude=%lf", asset->asset_id, latitude, longitude) < 0)
	{
		return NULL;
	}

	struct smm_curl_res_s *res = smm_connection_curl_retrieve_url (asset->conn, page, NULL, to_buffer, &buf);
	if (res == NULL)
	{
		free (page);
		return false;
	}
	if (!(res->success && res->httpcode == 200))
	{
		/* login and try again */
		smm_curl_res_free (res);
		free (page);
		return false;
	}
	free (page);

	if (res->content_type != NULL && strcmp (res->content_type, "application/json") == 0)
	{
		json_t *json_root;
		json_error_t json_error;

		json_root = json_loadb (buf.data, buf.bytes, 0, &json_error);
		if (json_root)
		{
			const char *url = NULL;
			uint64_t distance = 0;
			uint64_t length = 0;
			uint64_t sweep_width = 0;
			json_t *tmp = json_object_get (json_root, "object_url");
			if (tmp)
			{
				url = json_string_value (tmp);
			}
			tmp = json_object_get (json_root, "distance");
			if (tmp)
			{
				distance = json_integer_value (tmp);
			}
			tmp = json_object_get (json_root, "length");
			if (tmp)
			{
				length = json_integer_value (tmp);
			}
			tmp = json_object_get (json_root, "sweep_width");
			if (tmp)
			{
				sweep_width = json_integer_value (tmp);
			}
			search = smm_search_create (asset, url, length, distance, sweep_width);
			json_decref (json_root);
		}
	}
	smm_curl_res_free (res);
	free (buf.data);
	return search;
}


src/smm-asset-internal.h
#pragma once

/**
 * smm-asset-internal.h, Internal structures and headers for libsmm-asset
 *
 * Copyright 2019 Canterbury Air Patrol Incorporated
 *
 * This file is part of libsmm-asset
 *
 * libsmm-asset is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "smm-asset.h"

#include <pthread.h>
#include <stdbool.h>

#include <curl/curl.h>

extern bool smm_debug;
#define DEBUG(...) do \
	{ \
		if (smm_debug) \
		{ \
			printf ("%s:%i ", __func__, __LINE__);\
			printf(__VA_ARGS__); \
		} \
	} \
	while (0)

struct smm_connection_s
{
	char *host;
	char *user;
	char *pass;
	smm_connection_status state;
	CURL *curl;
	char *csrfmiddlewaretoken;
	pthread_mutex_t lock;
};

struct smm_asset_s
{
	smm_connection conn;
	char *name;
	char *type;
	int asset_id;
	int asset_type_id;
	smm_asset_command last_command;
	float last_command_lat;
	float last_command_lon;
};

struct smm_search_s
{
	smm_asset asset;
	char *url;
	uint64_t distance;
	uint64_t length;
	uint32_t sweep_width;
};

struct smm_curl_res_s
{
	bool success;
	long httpcode;
	char *full_uri;
	char *redirect_url;
	char *content_type;
};

struct buffer_s
{
	char *data;
	size_t bytes;
};

size_t to_buffer (char *ptr, size_t size, size_t nmemb, void *userdata);

void smm_curl_res_free (struct smm_curl_res_s *);
struct smm_curl_res_s *smm_connection_curl_retrieve_url (smm_connection conn, const char *path, const char *post_data,
							 size_t (*write_func) (char *ptr, size_t size, size_t nmemb, void *userdata), void *write_data);
bool smm_connection_login (smm_connection connection);

smm_asset smm_asset_create (smm_connection connection, const char *name, const char *type, int asset_id, int asset_type_id);
void smm_asset_free_asset (smm_asset assets);


src/smm-asset-curl.c
/**
 * smm-asset-curl.c, Use curl to communicate with the SMM server.
 *
 * Copyright 2019 Canterbury Air Patrol Incorporated
 *
 * This file is part of libsmm-asset
 *
 * libsmm-asset is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "config.h"

#include "smm-asset.h"
#include "smm-asset-internal.h"

#include <pthread.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_TIDY_H
#include <tidy.h>
#include <tidybuffio.h>
#elif HAVE_TIDY_TIDY_H
#include <tidy/tidy.h>
#include <tidy/tidybuffio.h>
#else
#error No tidy header(s)
#endif

void
smm_curl_res_free (struct smm_curl_res_s *res)
{
	if (res)
	{
		free (res->full_uri);
		free (res->redirect_url);
		free (res->content_type);
		free (res);
	}
}

static size_t
eat_data (char *ptr __attribute__ ((unused)), size_t size, size_t nmemb, void *userdata __attribute__ ((unused)))
{
	return size * nmemb;
}

size_t
to_buffer (char *ptr, size_t size, size_t nmemb, void *userdata)
{
	size_t new_bytes = size * nmemb;
	struct buffer_s *buf = (struct buffer_s *) userdata;
	char *tmp = realloc (buf->data, buf->bytes + new_bytes + 1);
	if (tmp == NULL)
	{
		return 0;
	}
	buf->data = tmp;
	memcpy (&buf->data[buf->bytes], ptr, new_bytes);
	buf->bytes += new_bytes;
	buf->data[buf->bytes] = '\0';

	return new_bytes;
}

static struct smm_curl_res_s *
_smm_connection_curl_retrieve_url (smm_connection conn, const char *path, const char *post_data,
				   size_t (*write_func) (char *ptr, size_t size, size_t nmemb, void *userdata), void *write_data)
{
	CURL *curl;
	struct smm_curl_res_s *res = NULL;

	DEBUG ("(%p, %s, %s, %p)\n", (void *) conn, path, post_data, write_data);

	if (conn == NULL || path == NULL)
	{
		DEBUG ("conn or path is NULL\n");
		return NULL;
	}

	pthread_mutex_lock (&conn->lock);
	curl = conn->curl;
	if (curl == NULL)
	{
		DEBUG ("creating curl object\n");
		curl = curl_easy_init ();
		conn->curl = curl;
	}

	res = (struct smm_curl_res_s *) calloc (1, sizeof (struct smm_curl_res_s));

	if (asprintf (&res->full_uri, "%s%s", conn->host, path) < 0)
	{
		free (res);
		DEBUG ("failed to allocate full_uri");
		pthread_mutex_unlock (&conn->lock);
		return NULL;
	}

	curl_easy_setopt (curl, CURLOPT_FAILONERROR, true);
	curl_easy_setopt (curl, CURLOPT_SSL_VERIFYPEER, 0L);
	curl_easy_setopt (curl, CURLOPT_SSL_VERIFYHOST, 0L);
	curl_easy_setopt (curl, CURLOPT_COOKIEFILE, "");
	curl_easy_setopt (curl, CURLOPT_FOLLOWLOCATION, 0L);
	curl_easy_setopt (curl, CURLOPT_URL, res->full_uri);

	if (post_data)
	{
		curl_easy_setopt (curl, CURLOPT_REFERER, res->full_uri);
		curl_easy_setopt (curl, CURLOPT_POSTFIELDS, post_data);
		curl_easy_setopt (curl, CURLOPT_POST, 1);
	}
	else
	{
		curl_easy_setopt (curl, CURLOPT_REFERER, NULL);
		curl_easy_setopt (curl, CURLOPT_POSTFIELDS, NULL);
		curl_easy_setopt (curl, CURLOPT_POST, 0);
	}

	if (write_func)
	{
		curl_easy_setopt (curl, CURLOPT_WRITEFUNCTION, write_func);
		curl_easy_setopt (curl, CURLOPT_WRITEDATA, write_data);
	}
	else
	{
		curl_easy_setopt (curl, CURLOPT_WRITEFUNCTION, eat_data);
		curl_easy_setopt (curl, CURLOPT_WRITEDATA, NULL);
	}

	DEBUG ("fetching %s\n", res->full_uri);
	CURLcode cres = curl_easy_perform (curl);
	DEBUG ("curl returned %i\n", cres);
	res->success = (cres == CURLE_OK);

	curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &res->httpcode);
	DEBUG ("httpcode = %li\n", res->httpcode);
	switch (res->httpcode)
	{
		case 200:
		{
			char *ct = NULL;
			if (curl_easy_getinfo (curl, CURLINFO_CONTENT_TYPE, &ct) == CURLE_OK)
			{
				res->content_type = strdup (ct);
			}
		}
			break;
		case 301:
		case 302:
		case 303:
		{
			char *redirect_url = NULL;
			if (curl_easy_getinfo (curl, CURLINFO_REDIRECT_URL, &redirect_url) == CURLE_OK)
			{
				res->redirect_url = strdup (redirect_url);
			}
		}
			break;
	}

	/* Clear anything we set in the curl object */
	curl_easy_setopt (curl, CURLOPT_URL, NULL);
	curl_easy_setopt (curl, CURLOPT_REFERER, NULL);
	curl_easy_setopt (curl, CURLOPT_POSTFIELDS, NULL);
	curl_easy_setopt (curl, CURLOPT_POST, 0);
	curl_easy_setopt (curl, CURLOPT_WRITEFUNCTION, NULL);
	curl_easy_setopt (curl, CURLOPT_WRITEDATA, NULL);

	pthread_mutex_unlock(&conn->lock);

	DEBUG ("Done\n");

	return res;
}

static size_t
populate_tidy (char *ptr, size_t size, size_t nmemb, void *userdata)
{
	tidyBufAppend ((TidyBuffer *) userdata, ptr, size * nmemb);
	return size * nmemb;
}

static bool
_extract_csrfmiddlewaretoken (TidyDoc tdoc, TidyNode tnod, char **token)
{
	bool res = false;
	for (TidyNode child = tidyGetChild (tnod); child; child = tidyGetNext (child))
	{
		ctmbstr name = tidyNodeGetName (child);
		if (name)
		{
			if (strcmp (name, "input") == 0)
			{
				/* check the attributes */
				for (TidyAttr attr = tidyAttrFirst (child); attr; attr = tidyAttrNext (attr))
				{
					ctmbstr attrName = tidyAttrName (attr);
					if (strcmp (attrName, "name") == 0)
					{
						if (strcmp (tidyAttrValue (attr), "csrfmiddlewaretoken") == 0)
						{
							res = true;
						}
					}
					else if (strcmp (attrName, "value") == 0)
					{
						if (res)
						{
							*token = strdup (tidyAttrValue (attr));
							return true;
						}
					}
				}
			}
		}
		res = _extract_csrfmiddlewaretoken (tdoc, child, token);
		if (res)
		{
			return res;
		}
	}
	return res;
}



bool
smm_connection_login (smm_connection connection)
{
	bool res = false;
	TidyDoc tdoc;
	TidyBuffer docbuf = { 0 };

	tdoc = tidyCreate ();
	tidyOptSetBool (tdoc, TidyForceOutput, yes);
	tidyOptSetInt (tdoc, TidyWrapLen, 4096);
	tidyBufInit (&docbuf);

	/* Get the login page, so we can get the csrf cookie + token */
	struct smm_curl_res_s *res_get = smm_connection_curl_retrieve_url (connection, "/accounts/login/", NULL, populate_tidy, &docbuf);

	if (res_get && res_get->success && res_get->httpcode == 200)
	{
		tidyParseBuffer (tdoc, &docbuf);
		tidyCleanAndRepair (tdoc);

		/* find the input token with the csrfmiddlewaretoken */
		_extract_csrfmiddlewaretoken (tdoc, tidyGetRoot (tdoc), &connection->csrfmiddlewaretoken);

		if (connection->csrfmiddlewaretoken)
		{
			char *post_data = NULL;
			if (asprintf
			    (&post_data, "csrfmiddlewaretoken=%s&username=%s&password=%s", connection->csrfmiddlewaretoken, connection->user,
			     connection->pass) >= 0)
			{
				struct smm_curl_res_s *res_post = smm_connection_curl_retrieve_url (connection, "/accounts/login/", post_data, NULL, NULL);
				if (res_post && res_post->success && res_post->httpcode == 302)
				{
					res = true;
					connection->state = SMM_CONNECTION_CONNECTED;
				}
				else
				{
					connection->state = SMM_CONNECTION_AUTHENTICATION_FAILURE;
				}
				smm_curl_res_free (res_post);
				free (post_data);
			}
		}
	}
	else if (!res_get)
	{
		DEBUG ("No res object returned\n");
		connection->state = SMM_CONNECTION_NO_HOST_CONNECTION;
	}
	else
	{
		DEBUG ("success = %s, httpcode = %li\n", res_get->success ? "true" : "false", res_get->httpcode);
	}
	smm_curl_res_free (res_get);

	tidyBufFree (&docbuf);
	tidyRelease (tdoc);

	return res;
}

struct smm_curl_res_s *
smm_connection_curl_retrieve_url (smm_connection conn, const char *path, const char *post_data,
				  size_t (*write_func) (char *ptr, size_t size, size_t nmemb, void *userdata), void *write_data)
{
	bool retry = true;
	int retries = 0;
	struct smm_curl_res_s *res = _smm_connection_curl_retrieve_url (conn, path, post_data, write_func, write_data);

	while (retry && retries < 3 && res != NULL)
	{
		retry = false;
		retries++;
		if (res->success && res->httpcode == 302 && res->redirect_url)
		{
			DEBUG ("Got redirected to (%s) accessing %s\n", res->redirect_url, path);
			/* It's possible we need to upgrade to https */
			if (strncmp (conn->host, "https://", 8) != 0)
			{
				if (strncmp (res->redirect_url, "https://", 8) == 0)
				{
					/* Upgrade to https */
					DEBUG ("Upgrading to https\n");
					char *new_host = NULL;
					if (strncmp (conn->host, "http://", 7) == 0)
					{
						if (asprintf (&new_host, "https://%s", &conn->host[7]) < 0)
						{
							DEBUG ("Failed to create new host\n");
						}
					}
					else
					{
						if (asprintf (&new_host, "https://%s", conn->host) < 0)
						{
							DEBUG ("Failed to create new host\n");
						}
					}
					if (new_host)
					{
						free (conn->host);
						conn->host = new_host;
						retry = true;
					}
				}
			}
			else if (strstr (res->redirect_url, "accounts/login") != NULL)
			{
				DEBUG ("Login required\n");
				if (smm_connection_login (conn))
				{
					retry = true;
				}
			}
			else
			{
				DEBUG ("Redirected to %s\n", res->redirect_url);
			}
		}
		if (retry && retries < 3)
		{
			smm_curl_res_free (res);
			res = _smm_connection_curl_retrieve_url (conn, path, post_data, write_func, write_data);
		}
	}

	return res;
}
